modules/contact/src/actions/ContactAction.C

```diff
@@ -1047,6 +1061,149 @@ ContactAction::addNodeFaceContact()
   }
 }
 
+// Specialization for PointListAdaptor<MooseMesh::PeriodicNodeInfo>
+// Return node location from NodeBoundaryIDInfo pairs
+template <>
+inline const Point &
+PointListAdaptor<NodeBoundaryIDInfo>::getPoint(const NodeBoundaryIDInfo & item) const
+{
+  return *(item.first);
+}
+
+void
+ContactAction::createSidesetsFromNodeProximity()
+{
+  mooseInfo("The contact action is reading the list of boundaries and automatically pairs them "
+            "if distance among nodes is less than a specified distance.");
+
+  if (!_mesh)
+    mooseError("Failed to obtain mesh for automatically generating contact pairs.");
+
+  if (!_mesh->getMesh().is_serial())
+    paramError(
+        "automatic_pairing_boundaries",
+        "The generation of automatic contact pairs in the contact action requires a serial mesh.");
+
+  // Create automatic_pairing_boundaries_id
+  std::vector<BoundaryID> _automatic_pairing_boundaries_id;
+  for (const auto & sideset_name : _automatic_pairing_boundaries)
+    _automatic_pairing_boundaries_id.emplace_back(_mesh->getBoundaryID(sideset_name));
+
+  // Vector of pairs node-boundary id
+  std::vector<NodeBoundaryIDInfo> node_boundary_id_vector;
+
+  // Data structures to hold the boundary nodes
+  const ConstBndNodeRange & bnd_nodes = *_mesh->getBoundaryNodeRange();
+
+  for (const auto & bnode : bnd_nodes)
+  {
+    const BoundaryID boundary_id = bnode->_bnd_id;
+    const Node * node_ptr = bnode->_node;
+
+    // Make sure node is on a boundary chosen for contact mechanics
+    auto it = std::find(_automatic_pairing_boundaries_id.begin(),
+                        _automatic_pairing_boundaries_id.end(),
+                        boundary_id);
+
+    if (it != _automatic_pairing_boundaries_id.end())
+      node_boundary_id_vector.emplace_back(node_ptr, boundary_id);
+  }
+
+  // sort by boundary id
+  std::sort(node_boundary_id_vector.begin(),
+            node_boundary_id_vector.end(),
+            [](const NodeBoundaryIDInfo & first_pair, const NodeBoundaryIDInfo & second_pair)
+            { return first_pair.second > second_pair.second; });
+
+  // build kd-tree
+  using KDTreeType = nanoflann::KDTreeSingleIndexAdaptor<
+      nanoflann::L2_Simple_Adaptor<Real, PointListAdaptor<NodeBoundaryIDInfo>, Real, std::size_t>,
+      PointListAdaptor<NodeBoundaryIDInfo>,
+      LIBMESH_DIM,
+      std::size_t>;
+
+  // This parameter can be tuned. Others use '10'
+  const unsigned int max_leaf_size = 20;
```

Comment:
Should this be a parameter controllable in the input file?

Reply:
Not sure. We don't have it for the generation of the mortar segment mesh, which could potentially be much much more critical than here?

We can at some point gauge its influence by profiling industrial problems.